compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/calls/ConstructorProcessing.kt

```diff
@@ -289,3 +292,43 @@ private fun <F : FirFunction<F>> prepareSubstitutorForTypeAliasConstructors(
 
     return TypeAliasConstructorsSubstitutor(typeAliasSymbol, substitutor, copyFactory)
 }
+
+private fun prepareCopyConstructorForTypealiasNestedClass(
+    matchedSymbol: FirClassLikeSymbol<*>,
+    originalSymbol: FirConstructorSymbol,
+    session: FirSession,
+    bodyResolveComponents: BodyResolveComponents,
+): FirConstructorSymbol? {
+    // If the matched symbol is a type alias, and the expanded type is a nested class, e.g.,
+    //
+    //   class Outer {
+    //     inner class Inner
+    //   }
+    //   typealias OI = Outer.Inner
+    //   fun foo() { Outer().OI() }
+    //
+    // the chances are that `processor` belongs to [ScopeTowerLevel] (to resolve type aliases at top-level), which treats
+    // the explicit receiver (`Outer()`) as an extension receiver, whereas the constructor of the nested class may regard
+    // the same explicit receiver as a dispatch receiver (hence inconsistent receiver).
+    // Here, we add a copy of the nested class constructor, along with the outer type as an extension receiver, so that it
+    // can be seen as if resolving:
+    //
+    //   fun Outer.OI(): OI = ...
+    //
+    if (originalSymbol.callableId.classId?.isNestedClass == true && matchedSymbol is FirTypeAliasSymbol) {
```

Comment:
For such utility functions, it may be easier to handle pre-requisites first in a form `if (!necessaryCondition) return null`.
E.g., this one is only expected to work with type-aliases to nested classes, so it can be rewritten like:
```
if (matchedSymbol !is FirTypeAliasSymbol) return null
val innerTypeRef = originalSymbol.fir.returnTypeRef
val innerType = innerTypeRef.coneType.fullyExpandedType(session) as? ConeClassLikeType
// `outerType()` returns null for not-inner class
val outerType = bodyResolveComponents.outerClassManager.outerType(innerType) ?: return null
/* useful code */
```
This would help to avoid code nestedness that (subjectively) doesn't make code more readable.
But it's cosmetics and quite a subjective statement, so I don't insist here

Reply:
Ahhh, you're right; and I also prefer that style! No idea why I forgot that at the moment.

Not quite worth having a PR to clean up; if I have a chance to revisit around here with another issue, I may rewrite it.